bip-0360.mediawiki

```diff
@@ -167,389 +171,304 @@ HASH160 <ref name="hash160">Used by P2PKH, P2SH, and P2WPKH addresses, though no
-
-* 0x01 - Key type 0 - secp256k1
-* 0x02 - Key type 1 - FALCON-512
-* 0x04 - Key type 2 - CRYSTALS-Dilithium Level I
-* 0x08 - Key type 3 - SPHINCS+-128s
-* 0x10 - Unused
-* 0x20 - Unused
-* 0x40 - Unused
-* 0x80 - Reserved for if additional key types are added in the future
-
-Example key type bitmask using all supported key types:
-
-  0x01 | 0x02 | 0x04 | 0x08 = 0x0F
-
-==== Hash Commitment ====
-
-If there is only a single public key, the hash is computed as the HASH256 of the public key.
-
-In order to support multiple keys, as in the context of multisig or singlesig hybrid cryptography, the hash is
-computed as a commitment to a vector of public key hashes:
-
-1. Sort the public keys first by key type, then by public key value
-2. For each sorted public key, compute its HASH256
-3. Concatenate all the public key hashes in sorted order
-4. Prepend key type bitmask and threshold to the concatenated hashes
-5. Compute the HASH256 of the result
-
-For example with 4 public keys:
-
-  // First sort the public keys
-  sorted_pubkeys = sort_by_key_type_and_value([pubkey1, pubkey2, pubkey3, pubkey4])
-
-  // Then compute hashes of sorted keys
-  h1 = HASH256(sorted_pubkeys[0])
-  h2 = HASH256(sorted_pubkeys[1])
-  h3 = HASH256(sorted_pubkeys[2])
-  h4 = HASH256(sorted_pubkeys[3])
+P2QRH (Pay to Quantum Resistant Hash) is a new output type that commits to the root of a tapleaf merkle tree. It is functionally
+identical to a P2TR (Pay to Taproot) output with no quantum vulnerable key-spend path. The lack of a key-spend path in P2QRH in a simple
+output type with no internal key and key tweak. Instead a P2QRH output is just the 32 byte root of the tapleaf merkle tree defined
+as in [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki BIP-341].
 
-  // Concatenate all hashes
-  concatenated = h1 <nowiki>||</nowiki> h2 <nowiki>||</nowiki> h3 <nowiki>||</nowiki> h4
+[[File:bip-0360/merkletree.png|center|550px|thumb|]]
 
-  commitment = key_type_bitmask <nowiki>||</nowiki> threshold <nowiki>||</nowiki> concatenated
+To construct a P2QRH output we follow the same process as [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki BIP-341]
+to compute the tapscript merkle root. However, instead of the root of the Merkle tree being hashed together with the internal
+key in P2QRH the root is hashed by itself using the tag "QuantumRoot".
 
-  hash = HASH256(commitment)
+<source>
+D = tagged_hash("TapLeaf", bytes([leaf_version]) + ser_script(script))
+CD = tagged_hash("TapBranch", C + D)
+CDE = tagged_hash("TapBranch", E + CD)
+ABCDE = tagged_hash("TapBranch", AB + CDE)
+Root = tagged_hash("QuantumRoot", ABCDE)
+</source>
 
-With sort_by_key_type_and_value defined as:
+A P2QRH input witness provides the following:
 
-  def sort_by_key_type_and_value(pubkeys):
-      return sorted(pubkeys, key=lambda x: (x.key_type, x.public_key))
+<source>
+script input_0,
```

Comment:
Nit: I don't like this notation, for a brief second I thought this was referring to transaction inputs.

BIP-0341 refers to these items as the "initial stack" in one place, there might be a better term, too.